tests/behat/features/bootstrap/EmailContext.php

```diff
@@ -191,30 +132,53 @@ public function iRunTheDigestCron($frequency) {
    *
    * @Then /^(?:|I )should have an email with subject "([^"]*)" and "([^"]*)" in the body$/
    */
-  public function iShouldHaveAnEmailWithTitleAndBody($subject, $body) {
-    $found_email = $this->findSubjectAndBody($subject, [$body]);
-
-    if (!$found_email) {
-      throw new \Exception('There is no email with that subject and body.');
-    }
+  public function iShouldHaveAnEmailWithTitleAndBody(string $subject, string $body) : void {
+    $this->assertEmailWithSubjectAndBody($subject, [$body]);
   }
 
   /**
    * I read an email with multiple content.
    *
    * @Then I should have an email with subject :arg1 and in the content:
    */
-  public function iShouldHaveAnEmailWithTitleAndBodyMulti($subject, TableNode $table) {
+  public function iShouldHaveAnEmailWithTitleAndBodyMulti(string $subject, TableNode $table) : void {
     $body = [];
     $hash = $table->getHash();
     foreach ($hash as $row) {
       $body[] = $row['content'];
     }
 
-    $found_email = $this->findSubjectAndBody($subject, $body);
+    $this->assertEmailWithSubjectAndBody($subject, $body);
+  }
+
+  /**
+   * No emails have been sent.
+   *
+   * @Then no emails have been sent
+   */
+  public function noEmailsHaveBeenSent() : void {
+    $finder = $this->getSpooledEmails();
 
-    if (!$found_email) {
-      throw new \Exception('There is no email with that subject and body.');
+    $count = $finder->count();
+    if ($count !== 0) {
+      throw new \Exception("No email messages should have been sent, but found $count.");
+    }
+  }
+
+  /**
+   * I do not have an email with a specific subject.
+   *
+   * Can be used in case you don't want an email to be sent regardless of the
+   * body.
+   *
+   * @Then should not have an email with subject :subject
+   * @Then I should not have an email with subject :subject
+   */
+  public function iShouldNotHaveAnEmailWithSubject(string $subject) : void {
```

Comment:
Here also again it should be commented on the PR #3500 

Not sure exactly what you mean? This code is being executed while we run behat tests and I think it's good to have the exception inside the function itself to keep the context clear and stop the tests running as it would mean it already failed.

Reply:
I'm not sure I understand exactly either.

However the code in `tests/behat/features/bootstrap` are context classes for [our Behat tests](https://open-social.slite.com/app/docs/xcFmWURyXQAwRK/Behat). The function you reference defines a test step that we can use in our `.feature` files. The test step acts as an assertion to make sure some condition that should occur in the test is actually true. So we want it to throw an exception and abort our Behat test in case something is wrong, so that the developers see an error. The `findEmailsWithSubject` is the underlying function that actually returns a count of emails :)

Reply:
My idea is make it more reusable, so example, If I need in another feature a countable, I can use this function, but when it return an throwable error, it became be specific to be used here.

But, as it's a test, I guess we can go forward.

Reply:
Yeah I get that :D However we write our tests [using `Gherkin` in feature files](https://open-social.slite.com/app/docs/xcFmWURyXQAwRK/Behat), so we need something to have a `@given`, `@then`, or `@when` annotation to make it usable in our Gherkin files :D So in this case this is already the specific step and the function it calls is the function that returns the countable :)